compiler/rename/RnExpr.hs

```diff
@@ -324,10 +324,27 @@ rnExpr (RecordUpd { rupd_expr = expr, rupd_flds = Right flds})
               addErr $ text "For this to work enable NamedFieldPuns."
          ; (getField, fv_getField) <- lookupSyntaxNameLocalEnv (mkVarOcc "getField")
          ; (setField, fv_setField) <- lookupSyntaxNameLocalEnv (mkVarOcc "setField")
-         ; (e, fv_e) <- rnLExpr expr
+         ; let arg_lit = fsLit "arg"
+               arg_name = mkSystemVarName (mkVarOccUnique arg_lit) arg_lit
+               arg_syn_expr = mkRnSyntaxExpr arg_name
+               arg_pat = noLoc $ VarPat noExt $ noLoc arg_name
+               arg_var = syn_expr arg_syn_expr
+         ; (expr', fv_expr) <- rnLExpr expr
          ; (us, fv_us) <- rnHsUpdProjs flds
-         ; return ( mkRecordDotUpd getField setField e us
-                  , plusFVs [fv_getField, fv_setField, fv_e, fv_us] )
+         ; return $
+            if (hasNestedUpdates us)
+            then
+               -- We preceed the record dot update with a
+               --
+               -- 'case record_expr of
+               --    x -> setField ... x'
+               --
+               -- for better type inference in the case of nested updates.
+              let match = mkHsCaseAlt arg_pat (noLoc $ mkRecordDotUpd getField setField (noLoc arg_var) us)
+              in (HsCase noExt expr' (mkMatchGroup Generated [match])
```

Comment:
I think this doesn’t solve the performance issue but probably solves the type inference issue. Consider this example:
```
module Test where

data X = X with
  a : Int
  c : Y

data Y = Y with
  r : Int

f :  X -> X
f x = x with
  a = 1
  c.r = 0
```
Without your PR we turn this into
```
f x =   
  setField @"c"
    (setField @"r" 0 (getField @"c" (setField @"a" 1 x)))
    (setField @"a" 1 x)
```
Clearly not great, we duplicate `setField @"a" 1 x`.
With your PR however, things are not better: We turn this into
```
f x =  case x of
  x ->
  setField @"c"
    (setField @"r" 0 (getField @"c" (setField @"a" 1 x)))
    (setField @"a" 1 x)
```
which is even worse wtr to performance.
What we should be turning it into is the following:
```
f x =  case setField @"a" 1 x of
  x' -> setField @"c" (setField @"r" 0 (getField @"c" x')) x'
```
So I think the logic for this needs to move into `mkRecordDotUpd`.

Apologies I should have realized this much earlier.

Reply:
I'm generating a case binding for each update now according to your suggestion. Also optimizing the case of a non-nested record update. I think the result looks right but considering it's Friday night and this is fairly complex, I'll have another look on Monday.

Reply:
@cocreature this looks good to me now. we add a case binding for each update unless it's not a nested update. The tests in the Daml repo pass, in particular the tests for RecordUpdate.daml.